### DISCUSSION TITLE
Support for BL0906 energy metering chip

### DISCUSSION THREAD

**Discussion by rp2019433** [arendst/Tasmota#22167] [Q&A] [answered]
Athom has just released its 2 CH Energy Meter Made For ESPHome with ESPHome (https://esphome.io/components/sensor/bl0906).

I was wondering if Tasmota could support **BL0906** energy metering driver so this hardware can be flashed with Tasmota as well.

**Comment by arendst** [comment]
I'll have a look once found an english translated datasheet....

**Comment by rp2019433** [comment]
Hi @arendst .

Just translated BL0906 datasheet from Chinese to English. Maybe it can help.


[BL0906_V1.1_en.pdf](https://github.com/user-attachments/files/17061574/BL0906_V1.1_en.pdf)

**Comment by Jason2866** [comment]
The esphome implementation https://github.com/esphome/esphome/tree/dev/esphome/components/bl0906

**Comment by arendst** [comment, marked as answer]
Try latest dev branch using one of the following templates:

{"NAME":"Athom EM2","GPIO":[0,0,0,0,0,0,0,3200,11329,1,544,0,0,0,0,0,0,0,1,0,0,0],"FLAG":0,"BASE":1}
{"NAME":"Athom EM6","GPIO":[0,0,0,0,0,0,0,3200,11333,1,544,0,0,0,0,0,0,0,1,0,0,0],"FLAG":0,"BASE":1}

Recommended commands:
``EnergyCols <phases>`` - Change default 4 column GUI display to <phases> columns
``VoltRes 1`` - Change none to 1 decimal display
``FreqRes 1`` - Change none to 1 decimal display
``WattRes 2`` - Change none to 2 decimals display
``SetOption21 1`` - Display voltage
``SetOption129 1`` - Display energy for each phase instead of single sum
``SetOption150 1`` - Display no common voltage/frequency

![image](https://github.com/user-attachments/assets/831f1523-96dd-4ea5-83fd-d08c03b1a427)

**Comment by rp2019433** [comment]
Using Tasmota dev version  14.2.0.6 with Athom EM2.

I'm running some tests, but everything is working good so far. 

![image](https://github.com/user-attachments/assets/a2b02e0b-c08a-416a-bf93-119df4f17ad0)


Since there is a button on GPIO9, just changed template to consider it.

`{"NAME":"Athom EM2","GPIO":[0,0,0,0,0,0,0,3200,11329,32,544,0,0,0,0,0,0,0,1,0,0,0],"FLAG":0,"BASE":1}`

**Reply by arendst** [reply]
Thx for letting me know.
Yes there is a button but for Tasmota it doesn't do anything hence I didn't add it to the template.

**Reply by rp2019433** [reply]
> Thx for letting me know. Yes there is a button but for Tasmota it doesn't do anything hence I didn't add it to the template.

Thank you for adding support for Athom EM2 and EM6 (BL0906) on Tasmota.

**Comment by rp2019433** [comment]
When using Web Requests to gather sensors information,  I noticed temperature is referencing BL0909 and not BL0906, as JSON below. Can you verify it, please?

http://< ip >/cm?cmnd=status%2010

`{
StatusSNS: {
Time: "2024-09-29T11:42:45",
ENERGY: {
TotalStartTime: "2024-09-28T15:03:35",
Total: [
1.942,
1.97
],
Yesterday: [
0.599,
0.61
],
Today: [
1.343,
1.36
],
Power: [
133.41,
135.13
],
ApparentPower: [
158.58,
159.59
],
ReactivePower: [
85.7,
84.9
],
Factor: [
0.84,
0.85
],
Frequency: 60,
Voltage: 219.7,
Current: [
0.722,
0.726
],
BL0909: {
Temperature: 36.3
}
},
TempUnit: "C"
}
}`

**Reply by sfromis** [reply]
Yeah, "BL0909" is hardcoded in the driver. Typo?
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xnrg_energy/xnrg_24_bl0906.ino#L463

**Reply by arendst** [reply]
Yep. Type. Will fix.

**Reply by arendst** [reply]
Fixed.

**Comment by rp2019433** [comment]
Couldn't find on docs. 
When a energy meter device has multiple channels as Athom EM2 or EM6, is it possible to add a friendly name (or rename) O1, O2, etc so on Web interface, Web Requests and MQTT user can see this information?

Something like this, renaming O1 and O2 to Kitchen and Lights:
![image](https://github.com/user-attachments/assets/501c95ee-4103-4205-8984-6051b71250cd)

And JSON Web Request with a new attribute ChannelName:

`
{
   "StatusSNS":{
      "Time":"2024-09-29T14:00:18",
      "ENERGY":{
         "TotalStartTime":"2024-09-28T15:03:35",
         "ChannelName":[
            "Kitchen",
            "Lights"
         ],
         "Total":[
            2.247,
            2.279
         ],
         "Yesterday":[
            0.599,
            0.61
         ],
         "Today":[
            1.647,
            1.669
         ],
         "Power":[
            145,
            146.73
         ],
         "ApparentPower":[
            181.89,
            183.02
         ],
         "ReactivePower":[
            109.8,
            109.4
         ],
         "Factor":[
            0.8,
            0.8
         ],
         "Frequency":60,
         "Voltage":223.1,
         "Current":[
            0.815,
            0.82
         ],
         "BL0909":{
            "Temperature":36.7
         }
      },
      "TempUnit":"C"
   }
}
`

**Reply by arendst** [reply]
Nope. That's a task for the front-end like Domoticz or HomeAssistant

**Comment by mikz** [comment]
Athom says they officially support only single phase measurements. Would it be possible to make 3 phase measurements  supported in Tasmota? This is a 6kW 3 phase electric heater on O1 to O3 (each with full 2kW on). I know the total Energy comes from bl0906, but the datasheet seemed like the chip could support something like that.
<img width="512" alt="image" src="https://github.com/user-attachments/assets/5b982232-7c78-4c47-bf6e-c22e32126505">

**Reply by arendst** [reply]
No way.  The chip supports only one voltage input so single phase it will be.

**Reply by RolandJonkers** [reply]
How do I add the 2 energy values O1 and O2 in domoticz?

**Reply by arendst** [reply]
Use rules to pick the correct value from the teleperiod message and command `DzSend` to send it to Domoticz.

See docs for rule examples.

**Comment by rp2019433** [comment]
Hi.

Still running tests for Athom EM6 and EM2 with Tasmota. Everything working fine and reliable. Again, thank you for supporting this driver on Tasmota.


There is a field named **Active Power Total** on Web interface. I don't see this information on JSON output for MQTT messages and Web requests. Could you please verify?

`{"StatusSNS":{"Time":"2024-11-04T10:33:54","ENERGY":{"TotalStartTime":"2024-10-30T15:38:44","Total":[0.000,0.000,46.902,0.000,0.000,0.000],"Yesterday":[0.000,0.000,10.860,0.000,0.000,0.000],"Today":[0.000,0.000,4.765,0.000,0.000,0.000],"Power":[0.0,0.0,-452.2,0.0,0.0,0.0],"ApparentPower":[0.0,0.0,585.6,0.0,0.0,0.0],"ReactivePower":[0.0,0.0,372.1,0.0,0.0,0.0],"Factor":[0.00,0.00,-0.77,0.00,0.00,0.00],"Frequency":60.0,"Voltage":221.4,"Current":[0.000,0.000,2.644,0.000,0.000,0.000],"BL0906":{"Temperature":35.4}},"TempUnit":"C"}}`

![image](https://github.com/user-attachments/assets/559ea3fc-aca2-4ac9-bcfa-0ceecaf3a696)




By the way, I was not able to find a command to enable or disable field Active Power Total. For Athom EM2 it's not being shown.


![image](https://github.com/user-attachments/assets/61b270e6-a5be-4e1b-9177-b74475aa19bf)

**Reply by sfromis** [reply]
To me, that "Active Power Total" looks like nothing more than a sum of the values from "Active Power", and therefore not really "needed" in the JSON payloads; not having any additional info.

Looks like the line is only generated when multi-phase and there a mix of phases without and with energy production (as in negative values present).

**Reply by rp2019433** [reply]
> To me, that "Active Power Total" looks like nothing more than a sum of the values from "Active Power", and therefore not really "needed" in the JSON payloads; not having any additional info.
> 
> Looks like the line is only generated when multi-phase and there a mix of phases without and with energy production (as in negative values present).

It makes sense. Just wanted to confirm if it could be a not desired behavior.

**Reply by Richie-Rogers** [reply]
Hi,
Just got one of the Athom 6-Channel Energy Meters secondhand.
I couldn't access it, so contacted Athom who supplied me with firmware to install.
Great, I thought, until I saw it was NOT Tasmota - I thought they supplied Tasmota for this product.
Ok, so I go to https://tasmota.github.io/install/
I can connect to the device (an ESP32-c3 with 4M), but the only Tasmota32 firmware that results in a working Tasmota UI is under Unofficial:
Tasmota32-c3 (2M)

None of the clamp inputs show up - even after I've applied the template from the blakadder site:
{"NAME":"Athom EM6","GPIO":[0,0,0,0,0,0,0,3200,11333,32,544,0,0,0,0,0,0,0,1,0,0,0],"FLAG":0,"BASE":1}

Has anyone got one of these working with Tasmota?

Thanks,
Richie

**Reply by sfromis** [reply]
If you go into Configuration/Template, are you seeing GPIO8 as "BL0906 Rx", and 6 in the dropdown to the right? If no, you lost it when even going near that restricted 2M binary, which is only meant for devices with that borderline unusable amount of flash.

As you are not giving any details, I can't see what made you think that the normal 4M binary (which you get if you just leave the selection to default, to automatically match chip architecture) did not work, or what you did to try it.

**Reply by Richie-Rogers** [reply]
Hi,
I was using the wrong binary files. I ended up using the ESP32-c3 factory bin and that worked perfectly.
I hadn't realised that was the one I was supposed to use - this is my first time using an ESP32.

Thanks,
Richie
